### PR TITLE
Add TS Typings and minor changes

### PR DESCRIPTION
**Pull request by Favna**
c8811a8 changes `Object.assign` to Object spread syntax. A bump in the ECMAversion for ESLint was required for it to recognize this. It's more or less opinionated, however Object.Assign() requires a polyfill in unsupported environments with Object spread merely requires transpiling which ultimately is less work for end-users. While doing this ESLint also informed of a missing semicolon on line 17 so yay, the linting is green again now.

---

8ac2813 adds TypeScript typings. Typings were generated by TSC by first rewriting the lib to TypeScript then emitting only declarations from there. I use a TS based version in my own bot so I only had to make minor adjustments to emit from. @types/node needs to be dependencies rather than devDependencies since the return type is a Readable which is part of NodeJS core.

I took the tsconfig and tslint config from Discord.JS proper assuming those were the configurations you guys would like to adhere to.

Please note that I'm also willing to change this PR so the lib is TS based entirely and we publish a `dist` folder with `index.js` and `index.d.ts` as transpiled by the TSC instead. I'll leave that up to the maintainers to say.

---

8055400 bumps the ESLint dependency to close #8 since @dependabot[bot] was saying there is an update available

### CODE REVIEW

**Comment by Favna** [comment]
Addition:

The aforementioned full TypeScript port as an option can be found here for review. Like I said before, maintainers call if that should be the way going forward as opposed to providing typings or not. 

https://github.com/Favna/ytdl-core-discord/tree/typescript-port

**Comment by amishshah** [comment]
Am I blind or are there no typings actually included in the PR? 👀

**Comment by Favna** [comment]
Damn you gitignore.

File: package.json [hunk 1]
```diff
@@ -1,10 +1,13 @@
 {
   "name": "ytdl-core-discord",
-  "version": "1.0.3",
+  "version": "1.0.4",
```
Comment (amishshah) [inline package.json]: Could you unbump this version for now?

Comment (Favna) [reply package.json]: done

**Comment by amishshah** [comment]
Sorry to be even more of a pain, but are the extra typescript dependencies necessary? I'm not really a typescript user but I have written typings for prism-media, and all I had to do there was specify the types directory in package.json.

**Comment by Favna** [comment]
I'm not exactly sure what you're referring to so I'll try to answer two of the options I am guessing.

If you're talking about the devDependency of Typescript, that's required for TSLint. TSLint is well.. the TypeScript linter. I used the same config that DiscordJS uses for that and its function here like there is to confirm the typings adhere to the [TypeScript typing guidelines (taken from the extended config description)](https://github.com/unional/typescript-guidelines/tree/master/pages/typings). (see https://github.com/discordjs/discord.js/blob/master/tsconfig.json and https://github.com/discordjs/discord.js/blob/master/tslint.json for example). I went on the assumption that due to the very close relation this lib has to DiscordJS (I mean... you're the owner of both libs...) it would be good to give it the same testing quality as DiscordJS has for as far as typings are concerned.

If you're talking about whether the typings are required at all, yes they are. TypeScript cannot magically infer typings from JavaScript and lacking a typings file throws an error as it simply doesn't know what to do with the library. 

image

**Comment by amishshah** [comment]
Thanks for the explanations, I was referring to the former point about TSLint, as the library is so small is it really necessary to install the lint dependencies for a single small file? I feel like we could get away with just having the typings file without any extra dependencies added.

**Comment by Favna** [comment]
Yes I suppose you're right. 

If I'd have to explain why I did it this way I guess I'd say it's because at work we work test driven so we have a lot of seemingly small stupid tests too but ye that's for a much much larger system.

**Comment by Favna** [comment]
Well you're call if I should remove that part then. Just lmk.

**Comment by amishshah** [comment]
That's great once you've removed it I'll merge, thanks again!

**Pull request merged**